examples/gadgets/simple/trace_dns/main.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	igjson "github.com/inspektor-gadget/inspektor-gadget/pkg/datasource/formatters/json"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	_ "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/localmanager"
+	ocihandler "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci-handler"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/simple"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/socketenricher"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+)
+
+func do() error {
+	const opPriority = 50000
+	myOperator := simple.New("myHandler", simple.OnInit(func(gadgetCtx operators.GadgetContext) error {
+		for _, d := range gadgetCtx.GetDataSources() {
+			jsonEncoder, _ := igjson.New(d)
+
+			d.Subscribe(func(source datasource.DataSource, data datasource.Data) error {
+				jsonOutput := jsonEncoder.Marshal(data)
+				fmt.Printf("%s\n", jsonOutput)
+				return nil
+			}, opPriority)
+		}
+		return nil
+	}))
+
+	// There are some gadgets, like trace_dns, trace_sni, trace_network that
+	// require the local manager (or kube manager) operators to work.
+	// TODO: How to mark them and why is it a hard requirement?
```

Comment:
**Why a hard requirement?**

The local/kube manager operator is required for all the socket filter programs because they need to be container-aware to run in each container's netns. All this is done using the AttachContainer/DetachContainer interfaces implemented by the networktracer and called by the local/kube manager operator. Similar reasoning for sched cls programs. They require the local/kube manager operator.

**How to mark?**

Currently, based on the program types (socket filter and sched cls), the ebpf operator request the local/kube manager operator to get activated using the `NeedContainerEvents` context variable (I'll open a PR on this soon) in the `Prepare` phase. Maybe in the `Start` phase, before actually running the gadget, we could check if the local/kube manager operator got actually activated or not. In case it wasn't return an error and request for it: `SocketFilter and SchedCLS programs require an operator that generate container creation/deletion events`, or so.

To document that, maybe the required operators can be automatically added to metadata file?

Reply:
IMO we shouldn't try to couple gadgets and operators and this coupling we have is technical debt. SocketFilter gadgets should capture events on the host if the local/kube manager aren't used, SchedCLS ones should be able to be attached to host's interfaces too (by using the --iface param).

I'd suggest to open an issue about it and link it on the example. WDYT?

Reply:
> SocketFilter gadgets should capture events on the host if the local/kube manager aren't used
>
But we should show host events only if user request for them with the `--host` flag, which is also managed by the local manager operator 😅. So, if developer doesn't add the local manager operator, thus they can't specify the want the host events. In this situation, what should we do? Simply output nothing or return an error because we know the gadget won't produce anything?

At this point, we should detach `--host` implementation from the local manager.

> SchedCLS ones should be able to be attached to host's interfaces too (by using the --iface param)
>
The advantage here is that the `--iface` flag is manged by the ebpf operator. But what to do if user doesn't specify `--iface`? Return an error?

> I'd suggest to open an issue about it and link it on the example. WDYT?
>
Yep, agree.

Reply:
Opened #2892